lib.py

```diff
@@ -1,167 +1,284 @@
 
+async def build_code(author_name: str, author_id: int, tmp_dir: str):
+    """
+    Designed to be used with a basic rust container. Run the container
+    with `cargo build` to build the code. Code is mounted in as a volume,
+    so that binaries are saved between build/run.
+    """
+    logger.info("Running container to build code for %s", author_id)
     try:
-        await loop.run_in_executor(
-            None,
-            functools.partial(doc.images.build, path="runner", tag=f"ferris-elf-{msg.author.id}"),
-        )
+        loop = asyncio.get_event_loop()
+        out = await loop.run_in_executor(None, functools.partial(
+            doc.containers.run,
+            settings.docker.container_ref,
+            "timeout --kill-after=5s 30s cargo build",
+            remove=True,
+            stdout=True,
+            mem_limit="8g",
+            # network_mode="none", # Want no-network, but it's downloading crates. :(
+            volumes={tmp_dir: {'bind': '/app', 'mode': 'rw'}}
+        ))
+        out = out.decode("utf-8")
+        logger.debug("Build container output: %s", out)
         return True
-    except docker.errors.BuildError as err:
-        logger.exception("Build Error while building for %s", msg.author.name)
-        e = ""
-        for chunk in err.build_log:
-            e += chunk.get("stream") or ""
-        await msg.reply(f"Error building benchmark: {err}")
+    except docker.errors.ContainerError:
+        logger.exception("Error in docker while building code")
         return False
-    finally:
-        await status.delete()
-
 
-async def run_image(msg, aoc_input):
-    logger.info("Running image/benchmark for %s", msg.author.name)
-    # input = ','.join([str(int(x)) for x in input])
-    status = await msg.reply("Running benchmark...")
-    loop = asyncio.get_event_loop()
-    try:
-        out = await loop.run_in_executor(
-            None,
-            functools.partial(
-                doc.containers.run,
-                f"ferris-elf-{msg.author.id}",
-                f"timeout 60 ./target/release/ferris-elf",
-                environment=dict(INPUT=aoc_input),
-                remove=True,
-                stdout=True,
-                mem_limit="24g",
-                network_mode="none",
-            ),
+def load_answers(cursor: Cursor, day: int, part: int):
+    """Load the expected answers for each input file."""
+    rows = cursor.execute(
+            "SELECT key, answer2 FROM solutions WHERE day = ? AND part = ?",
+            (day, part),
         )
-        out = out.decode("utf-8")
-        print(out)
-        return out
-    except docker.errors.ContainerError as err:
-        logger.exception("Error while running image for %s", msg.author.name)
-        await msg.reply(f"Error running benchmark: {err}")
-    finally:
-        await status.delete()
 
+    answer_map = {}
+    for r in rows:
+        (key, answer) = r
+        answer_map[key] = answer
 
```

Comment:
cant we just replace this with
```suggestion
    answer_map = dict(rows)
```
?

Reply:
Probably, but I actually don't know. The Cursor is an iterator, so that checks out. I don't know what type iterating over a Cursor yields (hopefully a Row?), and then there's the issue of how the `dict()` constructor interacts with whatever type the iterator returns.